### ISSUE TITLE
Make `consensus::encode` private

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2779]
The `consensus::encode` module can be considered a code organisation thing, the public API should probably just be `consensus` (ie., re-export all the relevant `encode` stuff).

Attempted in #2212 but that has gone stale and this is likely better attacked after #2756.

**Comment by Kixunil** [comment]
I don't think so, `consensus` is confusing. There are two consensus-related things: encoding and validation. They are mostly orthogonal, so having both in `consensus` submodule is weird. I'd prefer `consensus_encoding` for encoding and verification should be a method on `Transaction`.

**Comment by apoelstra** [comment]
I'm cool with that. I think we should also expose the transaction validation in a `consensus_validation` module so that people can look at that and see that our only consensus-capable code is on transactions, and that that is gated on the bitcoinconsensus feature.

But agreed that just `consensus` is ambiguous, and also that encoding and validation are different enough that they can be in top-level modules rather than submodules of "`consensus`".

**Comment by tcharding** [comment]
We have validation for transactions and scripts. Apart from that I'm on board.

**Comment by tcharding** [comment]
`consensus` is likely  moving to `primitives` so added to that milestone.

**Comment by Kixunil** [comment]
How? It should be removed.

**Comment by apoelstra** [comment]
I believe @tcharding means specifically the consensus encoding. Which I think we agreed elsewhere to "flatten" into `consensus_encoding` rather than having the `consensus::encoding` structure.

`consensus::params` should also move (this looks like it's independent and could maybe move first). I don't know that we have a plan for where it should move if the `consensus` module goes away.

`consensus::validation` I think we intended to rename to `consensus_validation`. Probably that should go in `primitives` too but I don't think it's particularly urgent and I don't feel particularly strongly about it.

**Comment by Kixunil** [comment]
> I think we agreed elsewhere to "flatten" into `consensus_encoding` rather than having the `consensus::encoding` structure.

Yes but it should go to its own crate to not hold up stabilizing `primitives`.

> `consensus::params` should also move

I've completely forgot we have it. Sounds like `consensus_params` or somewhere near `Network`.

> `consensus_validation`. Probably that should go in `primitives` too

Probably one day in form of `Transaction::prepare_validation()`

**Comment by apoelstra** [comment]
> Yes but it should go to its own crate to not hold up stabilizing `primitives`.

Hmmm, so we'd have a separate crate which depends on `primitives`, and which defines the encoding traits and implements them for all the types in `primitives`? How would we communicate this to users? I wonder if we should just leave the traits in rust-bitcoin for now.

I'm okay with this two-crate split temporarily, but eventually I definitely want the encoding traits to be in `primitives`. These traits are core to what it means to be a "primitive" and the `primitives` crate is pretty-much useless without them. But I understand if you want to pull it out for now so we can iterate a bit on the varint stuff and other things.

**Comment by Kixunil** [comment]
> so we'd have a separate crate which depends on `primitives`, and which defines the encoding traits and implements them for all the types in `primitives`?

Yes.

> How would we communicate this to users?

Those using `bitcoin` will get it out of the box and those migrating away would have some kind of guide. I was already thinking of writing something long-form but only once we're close to stabilizing at least one crate.

> I wonder if we should just leave the traits in rust-bitcoin for now.

Certainly a possibility. A separate crate mainly gives people the option to reduce compile time if they don't need the other stuff.



> I definitely want the encoding traits to be in `primitives`. These traits are core to what it means to be a "primitive" and the `primitives` crate is pretty-much useless without them.

I think traits should be defined in another crate and the relationship inverted. This does break Clean Architecture but given your valid argument about uselessness of primitives without them I think it's fine.

> But I understand if you want to pull it out for now so we can iterate a bit on the varint stuff and other things.

Yes, multiple versions of the trait won't be that big problem if people only use them locally to encode stuff (typical usage) and not across the crates.

**Comment by apoelstra** [comment]
All good points. Agreed in general. But in this case, because (a) I'd like `primitives` to be a self-contained "here are the consensus types and how to encode them" crate, and (b) the encoding/decoding stuff will change basically never (because consensus changes basically never), I still think we want to aim for pulling the traits into `primitives`.

Splitting them long-term feels like it's making our lives easier (letting us break the encoding traits because, as you say, it will rarely cause people pain) at the expense of making our dep tree larger and less discoverable. So for this reason I'm tempted to just keep them in rust-bitcoin until we're ready to pull them into `primitives`.

But I'm fine deprioritizing it for this initial `primitives` release.

**Comment by Kixunil** [comment]
> I still think we want to aim for pulling the traits into `primitives`.

I think we want them implemented for `units::Amount` though. Reexporting should be fine.

> So for this reason I'm tempted to just keep them in rust-bitcoin until we're ready to pull them into `primitives`

In the `push_decode` PR I've put the new traits into the new crate and kept the old API for backcompat (it can call through). This removes the need to do all impls at once or mess around with naming (`new_consensus_encoding`?).

**Comment by apoelstra** [comment]
> I think we want them implemented for `units::Amount` though. Reexporting should be fine.

Yeah, agreed, we want the traits on `Amount`.

To clarify: `primitives` needs to depend on `units`. So you are suggesting that `primitives` can impl the encoding/decoding traits on the `units` types and re-export the types themselves?

I think this is a further argument in favor of having the traits directly in `primitives`. It's easy for `primitives` to do this, but if the traits were in their own crate, then users of `primitives` **and** users of `units` would need to know about and pull in the extra crate.

> In the `push_decode` PR I've put the new traits into the new crate and kept the old API for backcompat (it can call through). This removes the need to do all impls at once or mess around with naming (`new_consensus_encoding`?).

Nice. So maybe one direction is that in the coming months we pull the "old" consensus traits into `primitives`, then when the new traits are ready, we can deprecate them but keep them around for a long time (until we need to do a major rev of `primitives` at least) and also pull the new traits into `primitives`. Since both traits are able to live side-by-side and even reuse some code.

Or maybe push_decode will be ready earlier and we can just start by pulling *that* into primitives and leaving the old stuff in rust-bitcoin until we're willing/able to delete it. I guess that depends on reviewer time more than anything.

**Comment by Kixunil** [comment]
> So you are suggesting that `primitives` can impl the encoding/decoding traits on the `units` types and re-export the types themselves?

I was suggesting to define the traits and common helpers (hex, serde...) in a completely separate crate which would be depended on by both `units` and `primitives` (and other crates) each implementing the traits for its types and `primitives` also reexporting the trait. This has an advantage that users can encode/decode things that are not `primitives` without depending on `primitives` but from practical standpoint, this is unlikely to be needed so I wouldn't be too opposed to an "unclean" solution of just folding it into `primitives`.

> then users of `primitives` **and** users of `units` would need to know about and pull in the extra crate.

`primitives` would depend on it unconditionally because it'll be needed to compute `Txid` anyway. (I have a hack in mind to workaround this in the meantime without exposing it as stable API.)

> So maybe one direction is that in the coming months we pull the "old" consensus traits into `primitives`

I'd rather not do this. I think that releasing stable `primitives` sooner is more valuable than not having a period of time when people will read "for now you need this other crate to encode them, we're working on stabilizing it here" in the doc.

**Comment by apoelstra** [comment]
> I'd rather not do this. I think that releasing stable `primitives` sooner is more valuable than not having a period of time when people will read "for now you need this other crate to encode them, we're working on stabilizing it here" in the doc.

Yeah, that's fair. Especially if we expect many/most people to just keep using this stuff through the rust-bitcoin facade so they won't even notice the crate splits other than having an easier time handling upgrades.

**Comment by apoelstra** [comment]
> I was suggesting to define the traits and common helpers (hex, serde...) in a completely separate crate which would be depended on by both `units` and `primitives` (and other crates) each implementing the traits for its types and `primitives` also reexporting the trait. This has an advantage that users can encode/decode things that are not `primitives` without depending on `primitives` but from practical standpoint, this is unlikely to be needed so I wouldn't be too opposed to an "unclean" solution of just folding it into `primitives`.

In addition to not being very useful, it also means that any breaking changes in the encoding traits turn into major releases of both `primitives` and `units`. So it doesn't help us too much as far as breakage isolation goes.

**Comment by Kixunil** [comment]
> any breaking changes in the encoding traits turn into major releases of both `primitives` and `units`

Interesting point. I hope it never happens (goodbye coroutines! :cry: ) but it's somewhat nice to have the impact reduced.

**Comment by tcharding** [comment]
Unless I'm missing something I don't see a strong argument for splitting the traits into a separate crate and we are loosely ok with putting them in `primitives` (which is what I had in mind). Am I understanding you both correctly?

**Comment by apoelstra** [comment]
@tcharding yes, but not in the first release of `primitives`. We'll move other stuff that we're more confident in first.

**Comment by tcharding** [comment]
Cool, to be painfully explicit, initially leave it in `bitcoin`.

**Comment by Kixunil** [comment]
One thing to note: the hack I mentioned is absolutely required to compute `Txid`. Unfortunately I don't remember all the details of the hack so I can't explain it here. I'll have to look into my notes (a bit later).

**Comment by apoelstra** [comment]
Can we implement encoding in `internals` though, when the encoding traits live in `bitcoin` so `internals` doesn't know about them?

This hack seems alright to me (assuming it works). I would also be okay with just manually implementing txid and wtxid in `Transaction` for now. We have plenty of unit tests about it.

**Comment by Kixunil** [comment]
We'll implement `Encoder` from `push_decode` and wrap it downsream.

**Comment by apoelstra** [comment]
Oh, cute. So we're secretly using `push_decode` but not in the public API.

I think we probably don't need to get `internals` involved? In `primitives` itself we can depend on the push-decoding crate but only use its traits internally on private newtypes, to expose `txid` (and `wtxid` .. and `to_hex`/`Display` and parsing and so on).

**Comment by apoelstra** [comment]
Lol, now I remember why in the plan @tcharding and I had in mind, we were moving the encoding traits to `primitives` as pretty-much the first step. We really do need encoding of some form to get a lot of basic functionality in `primitives`.

**Comment by Kixunil** [comment]
> I think we probably don't need to get `internals` involved?

That's for de-duplicating the impl. I don't want to maintain two complicated impls even for a short time.

**Comment by tcharding** [comment]
Flagging that I created a `primitives::consenus_validation` module in https://github.com/rust-bitcoin/rust-bitcoin/pull/3039 and move the script stuff to it.

**Comment by Kixunil** [comment]
Please don't put validation to primitives. It's not a primitive and I don't think its API is stable enough and I'd rather not prolong stabilization of other things by including it.

**Comment by apoelstra** [comment]
Agreed we should delay dealing with validation for a while.

Also agreed that conceptually it probably shouldn't be in `primitives`, it may make sense to put it there (feature-gated) versus permanently requiring an extension trait to get `Transaction::validate`. Or maybe not. This is a somewhat obscure thing. Though it definitely does get usage because we get bug reports and feature requests about it.

**Comment by tcharding** [comment]
Another idea I had was to remove the `impl` blocks all together (on `Script` and `Transaction`) and just leave the free standing functions (the `impl` blocks just call through to the free standing functions anyways).

**Comment by tcharding** [comment]
The old encoding traits are destined for the graveyard eventually now that we have the `consensus_encoding` crate, closing this.
